sources/engine/Xenko.Voxels/Voxels/Voxelization/VoxelVolumeProcessor.cs

```diff
@@ -0,0 +1,112 @@
+    {
+        private Dictionary<VoxelVolumeComponent, DataVoxelVolume> renderVoxelVolumes = new Dictionary<VoxelVolumeComponent, DataVoxelVolume>();
+        public Dictionary<VoxelVolumeComponent, ProcessedVoxelVolume> processedVoxelVolumes = new Dictionary<VoxelVolumeComponent, ProcessedVoxelVolume>();
+        bool isDirty;
+        SceneSystem sceneSystem;
+        GraphicsDevice graphicsDevice;
+        CommandList commandList;
+
+        public VisibilityGroup VisibilityGroup { get; set; }
+        public RenderGroup RenderGroup { get; set; }
+        protected override void OnSystemAdd()
+        {
+            base.OnSystemAdd();
+
+            VisibilityGroup.Tags.Set(VoxelRenderer.CurrentRenderVoxelVolumes, renderVoxelVolumes);
+            VisibilityGroup.Tags.Set(VoxelRenderer.CurrentProcessedVoxelVolumes, processedVoxelVolumes);
+            VisibilityGroup.Tags.Set(VoxelRenderFeature.CurrentProcessedVoxelVolumes, processedVoxelVolumes);
+            sceneSystem = Services.GetService<SceneSystem>();
+            graphicsDevice = Services.GetService<IGraphicsDeviceService>().GraphicsDevice;
+            commandList = Services.GetService<CommandList>();
+        }
+
+        public override void Draw(RenderContext context)
+        {
+            RegenerateVoxelVolumes();
+        }
+        
+        public ProcessedVoxelVolume GetProcessedVolumeForComponent(VoxelVolumeComponent component)
+        {
+            if (!processedVoxelVolumes.TryGetValue(component, out var data))
+                return null;
+            return data;
+        }
+        public DataVoxelVolume GetRenderVolumeForComponent(VoxelVolumeComponent component)
+        {
+            if (!renderVoxelVolumes.TryGetValue(component, out var data))
+                return null;
+            return data;
+        }
+
+        protected override void OnEntityComponentAdding(Entity entity, VoxelVolumeComponent component, VoxelVolumeComponent data)
+        {
+            component.Changed += ComponentChanged;
+        }
+        protected override void OnEntityComponentRemoved(Entity entity, VoxelVolumeComponent component, VoxelVolumeComponent data)
+        {
+            component.Changed -= ComponentChanged;
+        }
+        private void ComponentChanged(object sender, EventArgs eventArgs)
+        {
+            isDirty = true;
+        }
+
+        private void RegenerateVoxelVolumes()
+        {
+            renderVoxelVolumes.Clear();
+            processedVoxelVolumes.Clear();
+            foreach (var pair in ComponentDatas)
+            {
+                if (!pair.Key.Enabled)
+                    continue;
+
+                var volume = pair.Key;
+
+                DataVoxelVolume data;
+                    renderVoxelVolumes.Add(volume, data = new DataVoxelVolume());
+                processedVoxelVolumes.Add(volume, new ProcessedVoxelVolume());
```

Comment:
We probably want to avoid allocating every frame.
Note: I am fine to get this improved on later after the initial merge, which will probably happen soon.

Reply:
Could this be solved by simply reusing a field?